File: android/core/src/main/kotlin/com/demosten/srednabg/core/ZoneStatusColor.kt
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2026 SrednaBG Contributors
//
// SrednaBG — android / core

package com.demosten.srednabg.core

const val ZONE_COLOR_GREEN: Int = 0xFF66BB6A.toInt()
const val ZONE_COLOR_YELLOW: Int = 0xFFFDD835.toInt()
const val ZONE_COLOR_RED: Int = 0xFFEF5350.toInt()

/**
 * The colour of "no verdict" — used for [ZoneState.Unmeasured], where we know the
 * driver is inside an average-speed zone but never saw the entry. Green / amber /
 * red *is* the verdict, so rendering any of them there would claim knowledge we
 * do not have.
 */
const val ZONE_COLOR_NEUTRAL: Int = 0xFF9E9E9E.toInt()

/**
 * Status traffic light for the in-zone HUD:
 * - **red** ([ZONE_COLOR_RED]) when the running average is already over the limit;
 * - **amber** ([ZONE_COLOR_YELLOW]) when the current instantaneous speed exceeds
 *   the **car** limit (recoverable — the average is still fine, ease off);
 * - **green** ([ZONE_COLOR_GREEN]) otherwise.
 *
 * The amber tier is deliberately keyed off `zone.speedLimits.car`, **not** the
 * vehicle-type-resolved limit, so the colour matches byte-for-byte across the
 * SwiftUI / UIKit / Compose surfaces (see ISSUE-002 — the iOS port documents the
 * same choice). Do **not** "fix" this to the per-vehicle limit without changing
 * both platforms together, or the traffic-light colour will diverge. Pass the
 * vehicle-aware limit upstream if per-vehicle colouring is ever wanted.
 */
fun zoneStatusColor(state: ZoneState.InZone, currentSpeedKmh: Double?): Int {
    return when {
        state.speedStatus.isOverLimit -> ZONE_COLOR_RED
        currentSpeedKmh != null && currentSpeedKmh > state.zone.speedLimits.car -> ZONE_COLOR_YELLOW
        else -> ZONE_COLOR_GREEN
    }
}

File: ios/Packages/SrednaBGCore/Sources/SrednaBGCore/ZoneStatusColor.swift
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2026 SrednaBG Contributors
//
// SrednaBG — ios / SrednaBGCore

import Foundation

/// 0xAARRGGBB packed colors matching Android (`Int32` to fit signed Int range).
public let zoneColorGreen: Int32 = Int32(bitPattern: 0xFF66BB6A)
public let zoneColorYellow: Int32 = Int32(bitPattern: 0xFFFDD835)
public let zoneColorRed: Int32 = Int32(bitPattern: 0xFFEF5350)

/// The colour of "no verdict" — used for `ZoneState.Unmeasured`, where we know
/// the driver is inside an average-speed zone but never saw the entry. Green /
/// amber / red *is* the verdict, so rendering any of them there would claim
/// knowledge we do not have. Mirrors Android's `ZONE_COLOR_NEUTRAL`.
public let zoneColorNeutral: Int32 = Int32(bitPattern: 0xFF9E9E9E)

/// Status traffic light: red if the running average is over the limit, yellow
/// if the current instantaneous speed exceeds the limit (recoverable), green
/// otherwise. Uses the **car** limit deliberately — this matches Android. Pass
/// the vehicle-aware limit upstream if you need per-vehicle coloring.
public func zoneStatusColor(state: ZoneState.InZone, currentSpeedKmh: Double?) -> Int32 {
    if state.speedStatus.isOverLimit { return zoneColorRed }
    if let speed = currentSpeedKmh, speed > Double(state.zone.speedLimits.car) {
        return zoneColorYellow
    }
    return zoneColorGreen
}
